KEYCLOAK-4998: Adjust SAML attribute matching

skjolber opened the pull request:
Adjustments to AttributeToRoleMapper.

skjolber committed:
KEYCLOAK-4998: Adjust SAML attribute matching

hmlnarik commented on services/src/main/java/org/keycloak/broker/saml/mappers/AttributeToRoleMapper.java:
```diff
@@ -56,13 +57,13 @@
         property.setName(ATTRIBUTE_NAME);
         property.setLabel("Attribute Name");
-        property.setHelpText("Name of attribute to search for in assertion.  You can leave this blank and specify a friendly name instead.");
+        property.setHelpText("Name-attribute to search for in assertion Attribute elements. You can leave this blank and specify a friendly name instead.");
```
Please revert.

hmlnarik commented on services/src/main/java/org/keycloak/broker/saml/mappers/AttributeToRoleMapper.java:
```diff
@@ -56,13 +57,13 @@
         property.setName(ATTRIBUTE_FRIENDLY_NAME);
         property.setLabel("Friendly Name");
-        property.setHelpText("Friendly name of attribute to search for in assertion.  You can leave this blank and specify a name instead.");
+        property.setHelpText("FriendlyName-attribute to search for in assertion Attribute elements. You can leave this blank and specify a name instead.");
```
Please revert.

skjolber replied:
The help text should be improved, name vs friendly name is not as clear as it could be.

hmlnarik replied:
Please revert.

hmlnarik commented on services/src/main/java/org/keycloak/broker/saml/mappers/AttributeToRoleMapper.java:
```diff
@@ -115,21 +123,32 @@ public void importNewUser(KeycloakSession session, RealmModel realm, UserModel u
         }
     }
-
+    
```
Please revert

hmlnarik commented on services/src/main/java/org/keycloak/broker/saml/mappers/AttributeToRoleMapper.java:
```diff
@@ -81,6 +82,13 @@
 
     public static final String PROVIDER_ID = "saml-role-idp-mapper";
 
+    private static final String emptyToNull(String value) {
```
Use existing functions instead, e.g. org.keycloak.saml.common.util.StringUtil.isNotNull(String).

skjolber replied:
Your StringUtil.isNotNull() and StringUtil.isNullOrEmpty() do and do not trim whitespace, should I make them consistent as well?

hmlnarik replied:
No

hmlnarik commented on services/src/main/java/org/keycloak/broker/saml/mappers/AttributeToRoleMapper.java:
```diff
@@ -115,21 +123,32 @@ public void importNewUser(KeycloakSession session, RealmModel realm, UserModel u
+                }
                 for (Object val : attr.getAttributeValue()) {
-                    if (val.equals(desiredValue)) return true;
+                    if (val != null && val.toString().equals(desiredValue)) {
```
Test needs to be there specifically to show the need for "toString()" conversion.

skjolber replied:
This solution is suboptimal as the toString() is really a subsistute for specifying the desfiredValue type in the user-interface. However it is problably good enough, I'll add a test.

hmlnarik reviewed (changes requested):
Please add / update the relevant integration test case with appropriate test(s) that would fail before this commit and get fixed by this commit.

skjolber commented:
Reverted some equality check changes and added links to the corresponding parser locations; while the XML Schema is open for various object types (like Boolean, Integer etc) naturally mapped to corresponding java types, values are now parsed to Strings, and never as null, so it is sufficient to reflect this in the value comparison.

skjolber commented on saml-core/src/main/java/org/keycloak/saml/common/util/StringUtil.java:
```diff
@@ -58,7 +58,21 @@ public static boolean isNotNull(String str) {
+     * @return null if value was null or length was zero after whitespace trimming.
+     */
 
+    public static final String nullOrEmptyOrWhitespaceToNull(String value) {
```
Your current StringUtil methods do not agree on what  'empty' is. Updating docs instead.

hmlnarik replied:
Please refrain from refactoring irrelevant parts of code. These are scheduled for another task. Kindly remove these changes.

skjolber replied:
The nullOrEmptyOrWhitespaceToNull is removed.

hmlnarik commented on services/src/main/java/org/keycloak/broker/saml/mappers/AttributeToRoleMapper.java:
```diff
@@ -17,6 +17,7 @@
 
 package org.keycloak.broker.saml.mappers;
 
+import org.jboss.logging.processor.util.Objects;
```
Please don't use Objects from org.jboss.logging. The original comment was referring to Objects.equals from standard java.util package

hmlnarik commented on services/src/main/java/org/keycloak/broker/saml/mappers/AttributeToRoleMapper.java:
```diff
@@ -115,19 +117,31 @@ public void importNewUser(KeycloakSession session, RealmModel realm, UserModel u
                 AttributeType attr = choice.getAttribute();
-                if (name != null && !name.equals(attr.getName())) continue;
-                if (friendly != null && !name.equals(attr.getFriendlyName())) continue;
+                if (name != null && !Objects.areEqual(name, attr.getName())) continue;
```
Keep the amount of changes minimal. Update the original condition if necessary, use Objects.equals instead of areEqual. Use e.g. 
```java
if (name != null && ! Objects.equals(name, attr.getName())) continue;
```

hmlnarik reviewed (changes requested):
The original request is still valid: Please add / update the relevant integration test case with appropriate test(s) that would fail before this commit and get fixed by this commit.

Other comments inline.

skjolber commented:
The PR build previously failed on some unrelated issue, which indicates there is some sensitivity to timing or available ports or something like that in one of the unit tests. The fix was triggering another build.

hmlnarik commented:
@skjolber The PR as it is now cannot be merged. The lack of tests is blocker for this PR to get in, as are the other inline comments.

skjolber commented:
@hmlnarik the comments/javadoc are related to understanding the method which has a bug. I'll look into adding some tests.

hmlnarik commented:
@skjolber I seem that I should have written "as are the other inline comments on GitHub to the changes in the PR". No Javadoc/Java comment was meant.

hmlnarik commented:
It seems that there is no intent to address the review comment requesting tests. Without tests (that were asked for continuously since the first review), this PR will be rejected.

@skjolber Are you planning to write the tests?

skjolber commented:
Yes. However I am on holiday for a few weeks, som progress is slow.

skjolber commented:
I have added a unit test (some weeks ago), which should break with the previous code.

hmlnarik commented:
@skjolber The test requested above is not a unit test but an integration test. Unit tests are accepted only in very exceptional cases, not in this PR. Please add this test into testsuite/integration-arquillian/base/tests project.

stianst commented:
@skjolber Firstly I'm no SAML expert so I'm not quite following what the proposed changes are fixing. Could you please explain in high level what the problem you are facing is? If you could show me how the SAML assertion looks like, what the protocol mapper config looks like, etc. that would be great.

skjolber commented:
@stianst SAML attribute XML elements can have 'name' and 'friendly-name' attributes. Keycloack have a filter construct which helps in mapping SAML attributes to user properties and such. 

This filter also has two optinal attributes, corresponding the above target attributes. Matching fails with nullpointer because of a simple typo, in the line below. 

` if (friendly != null && !name.equals(attr.getFriendlyName())) continue;`

should have been

` if (friendly != null && !friendly.equals(attr.getFriendlyName())) continue;`

If an integration test really is necessary, could you point me in the right direction?

stianst commented:
@skjolber With regards to testing it's just how we test things in Keycloak. We use functional/integration tests only and try to avoid unit tests. @hmlnarik do you have a suggestion to an existing integration test that can be extended?

hmlnarik commented:
{{SamlUserAttributeMapperTest}} could be a good start to create a test for {{AttributeToRoleMapper}}. It could be actually extended to test the roles similarly to what is already there for user attributes, the roles being retrieved e.g. from a token.

stianst commented:
Closing due to lack of feedback. Feel free to reopen if you can update the tests.
